storage/resolver.go

```diff
@@ -0,0 +1,40 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package storage
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/elastic/package-registry/packages"
+)
+
+type storageResolver struct {
+	artifactsPackagesURL url.URL
+	artifactsStaticURL   url.URL
+}
+
+func (resolver storageResolver) RedirectArtifactsHandler(w http.ResponseWriter, r *http.Request, p *packages.Package) {
+	nameVersionZip := fmt.Sprintf("%s-%s.zip", p.Name, p.Version)
+	artifactURL := resolver.artifactsPackagesURL.ResolveReference(&url.URL{Path: nameVersionZip})
+	http.Redirect(w, r, artifactURL.String(), http.StatusSeeOther)
+}
+
+func (resolver storageResolver) RedirectStaticHandler(w http.ResponseWriter, r *http.Request, p *packages.Package, resourcePath string) {
+	nameVersion := fmt.Sprintf("%s-%s/", p.Name, p.Version)
+	staticURL := resolver.artifactsStaticURL.
+		ResolveReference(&url.URL{Path: nameVersion}).
+		ResolveReference(&url.URL{Path: resourcePath})
+	http.Redirect(w, r, staticURL.String(), http.StatusSeeOther)
+}
+
+func (resolver storageResolver) RedirectSignaturesHandler(w http.ResponseWriter, r *http.Request, p *packages.Package) {
+	nameVersionSigZip := fmt.Sprintf("%s-%s.zip.sig", p.Name, p.Version)
+	signatureURL := resolver.artifactsPackagesURL.ResolveReference(&url.URL{Path: nameVersionSigZip})
+	http.Redirect(w, r, signatureURL.String(), http.StatusSeeOther)
```

Comment:
Maybe we should use the plain 301 (`http.StatusMovedPermanently`) or 308 (`http.StatusPermanentRedirect`). 303 (`http.StatusSeeOther`) seems to be used after PUT or PUSH operations (https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections#temporary_redirections) to change the method back to GET.

Reply:
301 might be permanently cached in browsers or intermittent proxy servers. If you decide to change endpoints, it might be hard to reroute traffic.

Maybe 302 (temporary)? WDYT?

Reply:
I suppose that this thread is not relevant anymore. We will reuse `cacheHeaders`.

Reply:
Umm, are files cached beyond the cache headers we are including? I think it'd be fine if browsers do some caching of these files, even if we change endpoints, the urls should end up redirecting to the same files.
If we wouldn't use redirects, we'd have the same risk of having resources cached.

Thinking on Permanent vs. Temporary, I'd say that here we are doing a Permanent redirect, we are not serving the files anymore from the registry but from the object storage, this is a "permanent" relocation.

Reply:
Fixed: `http.StatusMovedPermanently`